File: soup/soup.cpp
#include "base.hpp"

#if SOUP_STANDALONE || SOUP_CODE_INSPECTOR

#include <stdexcept>
#include <string>

using stdexception = std::exception;
using stdstring = std::string;

#if SOUP_CODE_INSPECTOR
#include "soup.h" // Dummy include, prevents IntelliSense from adding this
#else
typedef void (*void_func_t)();
#endif

#include "RaiiEmulator.hpp"

#include "base32.hpp"
#include "base40.hpp"
#include "base64.hpp"
#include "Canvas.hpp"
#include "cbResult.hpp"
#include "Chatbot.hpp"
#include "CidrSubnetInterface.hpp"
#include "country_names.hpp"
#include "DetachedScheduler.hpp"
#include "EstablishWebSocketConnectionTask.hpp"
#include "Hotp.hpp"
#include "HttpRequest.hpp"
#include "HttpRequestTask.hpp"
#include "InquiryLang.hpp"
#include "IpAddr.hpp"
#include "KeyGenId.hpp"
#include "MimeMessage.hpp"
#include "Mixed.hpp"
#include "Notifyable.hpp"
#include "Promise.hpp"
#include "QrCode.hpp"
#include "rsa.hpp"
#include "Server.hpp"
#include "ServerWebService.hpp"
#include "Totp.hpp"
#include "Uri.hpp"
#include "WebSocketMessage.hpp"
#include "Worker.hpp"
#include "YubikeyValidator.hpp"

using namespace soup;

static RaiiEmulator heap{};
static std::string ret_str_buf{};

[[nodiscard]] static const char* ret_str(std::string&& str)
{
	ret_str_buf = std::move(str);
	return ret_str_buf.c_str();
}

#define returnString(x) return ret_str(x);

// [global namespace]

SOUP_CEXPORT void beginScope()
{
	heap.beginScope();
}

SOUP_CEXPORT void endScope()
{
	heap.endScope();
}

SOUP_CEXPORT void broadenScope(void* inst)
{
	heap.broadenScope(inst);
}

SOUP_CEXPORT void endLifetime(void* inst)
{
	heap.free(inst);
}

SOUP_CEXPORT const char* tryCatch(void_func_t f)
{
	try
	{
		f();
	}
	catch (const std::exception& e)
	{
		returnString(e.what());
	}
	catch (...)
	{
		returnString("...");
	}
	return nullptr;
}

SOUP_CEXPORT void throwException(const char* msg)
{
	throw Exception(msg);
}

SOUP_CEXPORT const char* getCountryName(const char* country_code, const char* language_code)
{
	return soup::getCountryName(country_code, language_code);
}

// base32

SOUP_CEXPORT const char* base32_encode(const stdstring* x, bool pad)
{
	returnString(base32::encode(heap.get(x), pad));
}

SOUP_CEXPORT stdstring* base32_decode(const char* x)
{
	return heap.add(new std::string(base32::decode(x)));
}

// base40

SOUP_CEXPORT const char* base40_encode(const stdstring* x)
{
	returnString(base40::encode(heap.get(x)));
}

SOUP_CEXPORT stdstring* base40_decode(const char* x)
{
	return heap.add(new std::string(base40::decode(x)));
}

// base64

SOUP_CEXPORT const char* base64_encode(const stdstring* x)
{
	returnString(base64::encode(heap.get(x)));
}

// Bigint

SOUP_CEXPORT const char* Bigint_toString(const Bigint* x)
{
	returnString(x->toString());
}

// Canvas

SOUP_CEXPORT unsigned int Canvas_getWidth(const Canvas* x)
{
	return x->width;
}

SOUP_CEXPORT unsigned int Canvas_getHeight(const Canvas* x)
{
	return x->height;
}

SOUP_CEXPORT void Canvas_resizeNearestNeighbour(Canvas* x, unsigned int desired_width, unsigned int desired_height)
{
	x->resizeNearestNeighbour(desired_width, desired_height);
}

SOUP_CEXPORT const char* Canvas_toSvg(const Canvas* x, unsigned int scale)
{
	returnString(x->toSvg(scale));
}

SOUP_CEXPORT stdstring* Canvas_toNewPngString(const Canvas* x)
{
	return heap.add(x->toPng());
}

// cbResult

SOUP_CEXPORT const char* cbResult_getResponse(const cbResult* x)
{
	return heap.get(x).response.c_str();
}

SOUP_CEXPORT bool cbResult_isDelete(const cbResult* x)
{
	return heap.get(x).isDelete();
}

SOUP_CEXPORT int cbResult_getDeleteNum(const cbResult* x)
{
	return heap.get(x).getDeleteNum();
}

// Chatbot

SOUP_CEXPORT cbResult* Chatbot_process(const char* text)
{
	return heap.add(Chatbot::process(text));
}

// CidrSubnetInterface

SOUP_CEXPORT CidrSubnetInterface* CidrSubnetInterface_new(const char* range)
{
	return heap.add(CidrSubnetInterface::construct(range).release());
}

SOUP_CEXPORT bool CidrSubnetInterface_contains(CidrSubnetInterface* x, const char* ip_addr)
{
	return heap.get(x).contains(IpAddr(ip_addr));
}

// DetachedScheduler

SOUP_CEXPORT DetachedScheduler* DetachedScheduler_new()
{
#if SOUP_WASM
	return nullptr;
#else
	return heap.add(new DetachedScheduler());
#endif
}

SOUP_CEXPORT bool DetachedScheduler_isActive(DetachedScheduler* sched)
{
#if SOUP_WASM
	return false;
#else
	return heap.get(sched).isActive();
#endif
}

// EstablishWebSocketConnectionTask

SOUP_CEXPORT void* EstablishWebSocketConnectionTask_new(const char* uri)
{
#if SOUP_WASM
	return nullptr;
#else
	return heap.add(new SharedPtr<EstablishWebSocketConnectionTask>(new EstablishWebSocketConnectionTask(Uri(uri))));
#endif
}

SOUP_CEXPORT void* EstablishWebSocketConnectionTask_getSocket(void* x)
{
#if SOUP_WASM
	return nullptr;
#else
	return heap.add(new SharedPtr<WebSocketConnection>(heap.get<SharedPtr<EstablishWebSocketConnectionTask>>(x)->sock));
#endif
}

// Hotp

SOUP_CEXPORT stdstring* Hotp_generateSecret(size_t bytes)
{
	return heap.add(Hotp::generateSecret(bytes));
}

// HttpRequest

SOUP_CEXPORT HttpRequest* HttpRequest_new(const char* uri)
{
#if SOUP_WASM
	return nullptr;
#else
	return heap.add(new HttpRequest(Uri(uri)));
#endif
}

SOUP_CEXPORT void HttpRequest_setPayload(HttpRequest* x, const char* data)
{
#if !SOUP_WASM
	heap.get(x).setPayload(data);
#endif
}

SOUP_CEXPORT const char* HttpRequest_getPath(HttpRequest* x)
{
#if !SOUP_WASM
	return heap.get(x).path.c_str();
#else
	return nullptr;
#endif
}

// HttpRequestTask

SOUP_CEXPORT void* HttpRequestTask_newFromRequest(const HttpRequest* hr)
{
#if SOUP_WASM
	return nullptr;
#else
	return heap.add(new SharedPtr<HttpRequestTask>(new HttpRequestTask(HttpRequest(*hr))));
#endif
}

SOUP_CEXPORT void* HttpRequestTask_newFromUrl(const char* url)
{
	return heap.add(new SharedPtr<HttpRequestTask>(new HttpRequestTask(Uri(url))));
}

SOUP_CEXPORT const char* HttpRequestTask_getResponseBodyCStr(void* hrt)
{
	return heap.get<SharedPtr<HttpRequestTask>>(hrt)->result->body.c_str();
}

// InquiryLang

SOUP_CEXPORT Mixed* InquiryLang_execute(const char* x)
{
	if (auto res = InquiryLang::execute(x))
	{
		return res.release();
	}
	return nullptr;
}

SOUP_CEXPORT const char* InquiryLang_formatResultLine(const Mixed* x)
{
	returnString(InquiryLang::formatResultLine(heap.get(x)));
}

// KeyGenId

SOUP_CEXPORT KeyGenId* KeyGenId_newFromSeedsExport(unsigned int bits, const stdstring* str)
{
	return heap.add(new KeyGenId(bits, heap.get(str)));
}

SOUP_CEXPORT KeyGenId* KeyGenId_generate(unsigned int bits)
{
	return heap.add(KeyGenId::generate(bits));
}

SOUP_CEXPORT stdstring* KeyGenId_toSeedsExport(const KeyGenId* x)
{
	return heap.add(heap.get(x).toSeedsExport());
}

SOUP_CEXPORT RsaKeypair* KeyGenId_getKeypair(const KeyGenId* x)
{
	return heap.add(heap.get(x).getKeypair());
}

// MimeMessage

SOUP_CEXPORT void MimeMessage_addHeader(MimeMessage* x, const char* key, const char* value)
{
	heap.get(x).header_fields.emplace(key, value);
}

// Mixed

SOUP_CEXPORT bool Mixed_isCanvas(const Mixed* x)
{
	return heap.get(x).isCanvas();
}

SOUP_CEXPORT Canvas* Mixed_getCanvas(const Mixed* x)
{
	return &heap.get(x).getCanvas();
}

// PromiseBase

SOUP_CEXPORT bool PromiseBase_isPending(void* x)
{
	return heap.get<SharedPtr<PromiseBase>>(x)->isPending();
}

// Promise_WebSocketMessage

SOUP_CEXPORT const char* Promise_WebSocketMessage_getData(void* x)
{
	return heap.get<SharedPtr<Promise<WebSocketMessage>>>(x)->getResult().data.c_str();
}

// QrCode

SOUP_CEXPORT QrCode* QrCode_newFromText(const char* x)
{
	return heap.add(QrCode::encodeText(x));
}

SOUP_CEXPORT Canvas* QrCode_toNewCanvas(const QrCode* x, unsigned int border, bool black_bg)
{
	return heap.add(heap.get(x).toCanvas(border, black_bg));
}

// RsaKeypair

SOUP_CEXPORT const Bigint* RsaKeypair_getN(const RsaKeypair* x)
{
	return &heap.get(x).n;
}

SOUP_CEXPORT const Bigint* RsaKeypair_getP(const RsaKeypair* x)
{
	return &heap.get(x).p;
}

SOUP_CEXPORT const Bigint* RsaKeypair_getQ(const RsaKeypair* x)
{
	return &heap.get(x).q;
}

// Scheduler

SOUP_CEXPORT Scheduler* Scheduler_new()
{
	return heap.add(new Scheduler());
}

SOUP_CEXPORT void Scheduler_setDontMakeReusableSockets(Scheduler* sched)
{
#if !SOUP_WASM
	heap.get(sched).dont_make_reusable_sockets = true;
#endif
}

SOUP_CEXPORT void Scheduler_add(Scheduler* sched, void* spWorker)
{
	heap.get(sched).addWorker(SharedPtr<Worker>(*reinterpret_cast<SharedPtr<Worker>*>(spWorker)));
}

SOUP_CEXPORT bool Scheduler_shouldKeepRunning(Scheduler* sched)
{
	return heap.get(sched).shouldKeepRunning();
}

SOUP_CEXPORT void Scheduler_tick(Scheduler* sched)
{
	heap.get(sched).tick();
}

// Server

SOUP_CEXPORT bool Server_bind(Scheduler* serv, int port, ServerService* srv)
{
#if !SOUP_WASM
	static_assert(sizeof(Scheduler) == sizeof(Server));
	return heap.get<Server>(serv).bind(static_cast<uint16_t>(port), srv);
#else
	return false;
#endif
}

// ServerWebService

#if !SOUP_WASM
class ServerWebServiceFfi : public ServerWebService
{
public:
	HttpRequest* pending_request = nullptr;
	Socket* pending_request_socket;
private:
	Notifyable notifyable;

public:
	ServerWebServiceFfi()
		: ServerWebService(&onRequest)
	{
	}

private:
	static void onRequest(Socket& s, HttpRequest&& hr, ServerWebService& _self)
	{
		ServerWebServiceFfi& self = static_cast<ServerWebServiceFfi&>(_self);
		self.pending_request = &hr;
		self.pending_request_socket = &s;
		self.notifyable.wait();
	}

public:
	void setRequestHandled()
	{
		pending_request = nullptr;
		notifyable.notify_all();
	}
};
#endif

SOUP_CEXPORT ServerWebService* ServerWebService_new()
{
#if !SOUP_WASM
	return heap.add(new ServerWebServiceFfi());
#else
	return nullptr;
#endif
}

SOUP_CEXPORT bool ServerWebService_hasPendingRequest(ServerWebService* x)
{
#if !SOUP_WASM
	return heap.get<ServerWebServiceFfi>(x).pending_request != nullptr;
#else
	return false;
#endif
}

SOUP_CEXPORT const char* ServerWebService_getPendingRequestPath(ServerWebService* x)
{
#if !SOUP_WASM
	return heap.get<ServerWebServiceFfi>(x).pending_request->path.c_str();
#else
	return nullptr;
#endif
}

SOUP_CEXPORT void ServerWebService_ignoreRequest(ServerWebService* x)
{
#if !SOUP_WASM
	heap.get<ServerWebServiceFfi>(x).setRequestHandled();
#endif
}

SOUP_CEXPORT void ServerWebService_replyWithHtml(ServerWebService* x, const char* html)
{
#if !SOUP_WASM
	auto& srv = heap.get<ServerWebServiceFfi>(x);
	ServerWebService::sendHtml(*srv.pending_request_socket, html);
	srv.setRequestHandled();
#endif
}

SOUP_CEXPORT void ServerWebService_replyWith404(ServerWebService* x)
{
#if !SOUP_WASM
	auto& srv = heap.get<ServerWebServiceFfi>(x);
	ServerWebService::send404(*srv.pending_request_socket);
	srv.setRequestHandled();
#endif
}

// Totp

SOUP_CEXPORT Totp* Totp_new(const stdstring* secret)
{
	return heap.add(new Totp(heap.get(secret)));
}

SOUP_CEXPORT const char* Totp_getQrCodeUri(const Totp* x, const char* label, const char* issuer)
{
	returnString(heap.get(x).getQrCodeUri(label, issuer));
}

SOUP_CEXPORT int Totp_getValue(const Totp* x)
{
	return heap.get(x).getValue();
}

// WebSocketConnection

SOUP_CEXPORT void WebSocketConnection_wsSend(void* con, const char* text)
{
#if !SOUP_WASM
	return heap.get<SharedPtr<WebSocketConnection>>(con)->wsSend(text);
#endif
}

SOUP_CEXPORT void* WebSocketConnection_wsRecv(void* con)
{
#if SOUP_WASM
	return nullptr;
#else
	return heap.add(new SharedPtr(heap.get<SharedPtr<WebSocketConnection>>(con)->wsRecv()));
#endif
}

// Worker

SOUP_CEXPORT bool Worker_isWorkDone(void* x)
{
	return heap.get<SharedPtr<Worker>>(x)->isWorkDone();
}

// YubikeyValidator

SOUP_CEXPORT YubikeyValidator* YubikeyValidator_new(const char* id, const char* secret)
{
#if SOUP_WASM
	return nullptr;
#else
	return heap.add(new YubikeyValidator(id, secret));
#endif
}

SOUP_CEXPORT const char* YubikeyValidator_validate(const YubikeyValidator* x, const char* otp)
{
#if SOUP_WASM
	return "";
#else
	returnString(heap.get(x).validate(otp).device_id);
#endif
}

// std::exception

SOUP_CEXPORT const char* exception_what(const stdexception* x)
{
	returnString(x->what());
}

#endif

File: bindings/soup.js
(function()
{
	var soup_q = [];

	window.soup = {
		ready: false,
		use: function(f)
		{
			soup_q.push(f);
		}
	};

	var have_libsoup_js = function()
	{
		libsoup().then(function(soup)
		{
			soup.beginScope = soup.cwrap("beginScope", "void", []);
			soup.endScope = soup.cwrap("endScope", "void", []);
			soup.broadenScope = soup.cwrap("broadenScope", "void", ["number"]);
			soup.free = soup.cwrap("endLifetime", "void", ["number"]);
			soup.tryCatch = soup.cwrap("tryCatch", "string", ["function"]);
			soup.throwException = soup.cwrap("throwException", "void", ["string"]);
			soup.getCountryName = soup.cwrap("getCountryName", "string", ["string", "string"]);
			soup.base32 = {
				encode: soup.cwrap("base32_encode", "string", ["number", "boolean"]),
				decode: soup.cwrap("base32_decode", "number", ["string"]),
			};
			soup.base40 = {
				encode: soup.cwrap("base40_encode", "string", ["number"]),
				decode: soup.cwrap("base40_decode", "number", ["string"]),
			};
			soup.base64 = {
				encode: soup.cwrap("base64_encode", "string", ["number"]),
			};
			soup.Bigint = {
				toString: soup.cwrap("Bigint_toString", "string", ["number"]),
			};
			soup.Canvas = {
				getWidth: soup.cwrap("Canvas_getWidth", "number", ["number"]),
				getHeight: soup.cwrap("Canvas_getHeight", "number", ["number"]),
				resizeNearestNeighbour: soup.cwrap("Canvas_resizeNearestNeighbour", "void", ["number", "number", "number"]),
				toSvg: soup.cwrap("Canvas_toSvg", "string", ["number", "number"]),
				toNewPngString: soup.cwrap("Canvas_toNewPngString", "number", ["number"]),
			};
			soup.cbResult = {
				getResponse: soup.cwrap("cbResult_getResponse", "string", ["number"]),
				isDelete: soup.cwrap("cbResult_isDelete", "boolean", ["number"]),
				getDeleteNum: soup.cwrap("cbResult_getDeleteNum", "number", ["number"]),
			};
			soup.Chatbot = {
				process: soup.cwrap("Chatbot_process", "number", ["string"]),
			};
			soup.CidrSubnetInterface = {
				new: soup.cwrap("CidrSubnetInterface_new", "number", ["string"]),
				contains: soup.cwrap("CidrSubnetInterface_contains", "boolean", ["number", "string"]),
			};
			soup.DetachedScheduler = {
				new: soup.cwrap("DetachedScheduler_new", "number", []),
				isActive: soup.cwrap("DetachedScheduler_isActive", "boolean", ["number"]),
			};
			soup.EstablishWebSocketConnectionTask = {
				new: soup.cwrap("EstablishWebSocketConnectionTask_new", "number", ["string"]),
				getSocket: soup.cwrap("EstablishWebSocketConnectionTask_getSocket", "number", ["number"]),
			};
			soup.Hotp = {
				generateSecret: soup.cwrap("Hotp_generateSecret", "number", ["number"]),
			};
			soup.HttpRequest = {
				new: soup.cwrap("HttpRequest_new", "number", ["string"]),
				setPayload: soup.cwrap("HttpRequest_setPayload", "void", ["number", "string"]),
			};
			soup.HttpRequestTask = {
				newFromRequest: soup.cwrap("HttpRequestTask_newFromRequest", "number", ["number"]),
				newFromUrl: soup.cwrap("HttpRequestTask_newFromUrl", "number", ["string"]),
				getResponseBodyCStr: soup.cwrap("HttpRequestTask_getResponseBodyCStr", "string", ["number"]),
			};
			soup.InquiryLang = {
				execute: soup.cwrap("InquiryLang_execute", "number", ["string"]),
				formatResultLine: soup.cwrap("InquiryLang_formatResultLine", "string", ["number"]),
			};
			soup.KeyGenId = {
				newFromSeedsExport: soup.cwrap("KeyGenId_newFromSeedsExport", "number", ["number", "number"]),
				generate: soup.cwrap("KeyGenId_generate", "number", ["number"]),
				toSeedsExport: soup.cwrap("KeyGenId_toSeedsExport", "number", ["number"]),
				getKeypair: soup.cwrap("KeyGenId_getKeypair", "number", ["number"]),
			};
			soup.MimeMessage = {
				addHeader: soup.cwrap("MimeMessage_addHeader", "void", ["number", "string", "string"]),
			};
			soup.Mixed = {
				isCanvas: soup.cwrap("Mixed_isCanvas", "boolean", ["number"]),
				getCanvas: soup.cwrap("Mixed_getCanvas", "number", ["number"]),
			};
			soup.PromiseBase = {
				isPending: soup.cwrap("PromiseBase_isPending", "boolean", ["number"]),
			};
			soup.Promise_WebSocketMessage = {
				getData: soup.cwrap("Promise_WebSocketMessage_getData", "string", ["number"]),
			};
			soup.QrCode = {
				newFromText: soup.cwrap("QrCode_newFromText", "number", ["string"]),
				toNewCanvas: soup.cwrap("QrCode_toNewCanvas", "number", ["number", "number", "boolean"]),
			};
			soup.RsaKeypair = {
				getN: soup.cwrap("RsaKeypair_getN", "number", ["number"]),
				getP: soup.cwrap("RsaKeypair_getP", "number", ["number"]),
				getQ: soup.cwrap("RsaKeypair_getQ", "number", ["number"]),
			};
			soup.Scheduler = {
				new: soup.cwrap("Scheduler_new", "number", []),
				setDontMakeReusableSockets: soup.cwrap("Scheduler_setDontMakeReusableSockets", "void", ["number"]),
				add: soup.cwrap("Scheduler_add", "void", ["number", "number"]),
				shouldKeepRunning: soup.cwrap("Scheduler_shouldKeepRunning", "boolean", ["number"]),
				tick: soup.cwrap("Scheduler_tick", "void", ["number"]),
			};
			soup.Totp = {
				new: soup.cwrap("Totp_new", "number", ["number"]),
				getQrCodeUri: soup.cwrap("Totp_getQrCodeUri", "string", ["number", "string", "string"]),
				getValue: soup.cwrap("Totp_getValue", "number", ["number"]),
			};
			soup.WebSocketConnection = {
				wsSend: soup.cwrap("WebSocketConnection_wsSend", "void", ["number", "string"]),
				wsRecv: soup.cwrap("WebSocketConnection_wsRecv", "number", ["number"]),
			};
			soup.Worker = {
				isWorkDone: soup.cwrap("Worker_isWorkDone", "boolean", ["number"]),
			};
			soup.YubikeyValidator = {
				new: soup.cwrap("YubikeyValidator_new", "number", ["string", "string"]),
				validate: soup.cwrap("YubikeyValidator_validate", "string", ["number", "string"]),
			};
			soup.exception = {
				what: soup.cwrap("exception_what", "string", ["number"]),
			};

			soup.CppException = class CppException extends Error
			{
			};
			soup.fixErrorType = function(e)
			{
				if (typeof e == "number")
				{
					return new soup.CppException(soup.exception.what(e));
				}   
				return e;
			}
			soup.tryCatch = function(f)
			{
				try
				{
					f();
				}
				catch (e)
				{
					throw soup.fixErrorType(e);
				}
			};
			soup.scope = function(f)
			{
				soup.beginScope();
				try
				{
					f();
				}
				catch (e)
				{
					soup.endScope();
					throw soup.fixErrorType(e);
				}
				soup.endScope();
			};

			soup.Canvas.upscaleMultiply = function(c, f)
			{
				let w = soup.Canvas.getWidth(c);
				let h = soup.Canvas.getHeight(c);
				soup.Canvas.resizeNearestNeighbour(c, w * f, h * f);
			};

			soup.ready = true;
			soup.use = function(f)
			{
				f();
			};
			window.soup = soup;

			findElements();
			enableObserver();

			soup_q.forEach(function(f)
			{
				f();
			});
		});
	};

	if("libsoup" in window)
	{
		have_libsoup_js();
	}
	else
	{
		var script = document.createElement("script");
		script.src = document.currentScript.src.replace("/soup.js","/libsoup.js");
		script.onload = have_libsoup_js;
		document.currentScript.parentNode.appendChild(script);
	}

	function getContainer(elm)
	{
		if(elm.parentNode.className == "soup-container")
		{
			return elm.parentNode;
		}
		let div = document.createElement("div");
		div.className = "soup-container";
		elm.parentNode.insertBefore(div, elm);
		div.appendChild(elm);
		return div;
	}

	function pruneContainer(div)
	{
		let elm = div.removeChild(div.children[0]);
		div.innerHTML = "";
		div.appendChild(elm);
	}

	function findElements()
	{
		document.querySelectorAll("soup-qr-code").forEach(elm => {
			let div = getContainer(elm);
			let svg = div.querySelector("svg");
			let text = elm.textContent;
			if(svg === null || svg.getAttribute("alt") != text)
			{
				soup.scope(function()
				{
					let qr = soup.QrCode.newFromText(text);
					let c = soup.QrCode.toNewCanvas(qr, 4, elm.hasAttribute("inverted"));
					soup.Canvas.upscaleMultiply(c, 4);
					let ps = soup.Canvas.toNewPngString(c);
					let pb = soup.base64.encode(ps);

					elm.style.display = "none";
					pruneContainer(div);
					let img = document.createElement("img");
					img.src = "data:image/png;base64," + pb;
					img.setAttribute("alt", text);
					div.appendChild(img);
				});
			}
		});
	}

	var observer = new MutationObserver(function(mutations)
	{
		disableObserver();
		findElements();
		enableObserver();
	});

	function enableObserver()
	{
		observer.observe(document, {attributes: true, childList: true, characterData: true, subtree: true });
	}

	function disableObserver()
	{
		observer.disconnect();
	}
})();
